### PR TITLE
many: allow mimics in parents of layout

### PR DESCRIPTION
**Pull request by zyga**
NOTE: This patch is just for testing, I will split it into smaller
chunks.

This patch fixes the layout system to allow creation of a writable mimic
in any reasonable parent of the affected layout source or target path.
In addition when creating the source path for a layout mount rule we now
allow poking holes in read-only places since that will be consumed
internally by the snap.

When the layout for /usr/lib/foo/bar/ is requested and /usr/lib/foo/
doesn't exist we really need a rule that allows any path of
/usr/lib/foo/bar/ not to exist.

To illustrate the issue, let's say that /snap/mysnap/123/a/b/c/ is laid
out on top of /usr/a/b/c/. Normally we would allow a writable mimic to
be constructed at /usr/share/a/b/ so that c may be created. As is
evident by this trivial example it is unlikely that /usr/share/a exists
and it also may need to be created.

This is achieved with the "chop tree" of apparmor alternation
expressions with a fixed-size "trunk" of non-optional elements. In this
case it allows the mimic to be created in /usr/{,share/{a/,{b/}}} since
we only assume /usr exists. The other side of the layout will have
similar rule that allows a mimic to be constructed over
/snap/mysnap/123/{,a/{,b/}} since we assume that /snap/mysnap/123 exists
as well.

This makes it possible to express natural layouts without understanding
the implementation or its limitations. The same patch adjusts the
content interface that uses some of the same helpers.

Signed-off-by: Zygmunt Krynicki <zygmunt.krynicki@canonical.com>

### CODE REVIEW

**Comment by zyga** [comment]
@jdstrand hey

Please have a look at this PR. This shows an issue with layout design that we ran into while trying to use layouts in practice. One simple suggestion from @niemeyer is to just allow the parents. If we need a layout for `/usr/lib/foo/bar/` we keep the layout permissions as-is but change the writable mimic permission to `/usr/{,lib/{,foo{,bar/}}}`

**Comment by jdstrand** [comment]
"Please have a look at this PR. This shows an issue with layout design that we ran into while trying to use layouts in practice. One simple suggestion from @niemeyer is to just allow the parents. If we need a layout for /usr/lib/foo/bar/ we keep the layout permissions as-is but change the writable mimic permission to /usr/{,lib/{,foo{,bar/}}}"

This concept seems fine-- the snap can just as easily request /usr/lib as the layout as it can /usr/lib/foo/bar so there is no change in terms of security or design.

File: interfaces/apparmor/spec_test.go [hunk 1]
```diff
@@ -280,3 +280,97 @@ func (s *specSuite) TestApparmorSnippetsFromLayout(c *C) {
...
+
+const snapWithTrickyLayout = `
+name: tricky
+version: 0
+apps:
+  tricky:
+    command: tricky
+layout:
+  /usr/lib/tcltk/x86_64-linux-gnu/tk8.5:
+    bind: $SNAP/usr/lib/tcltk/x86_64-linux-gnu/tk8.5
```
Comment (stolowski) [inline interfaces/apparmor/spec_test.go]: :smiley_cat:

**Comment by zyga** [comment]
I'm still working on the details of the pattern. I have something that should work that I need to polish now. I will publish it soon. Please don't land this yet.

**Comment by zyga** [comment]
I'm marking this as blocked. This branch is just for testing and backup. I will split this into smaller chunks and propose it separately.

**Comment by zyga** [comment]
I will close this as most of the branch has been proposed separately now. This was mainly my backup and explore workspace.

**Pull request closed without merging**